src/datasets/arrow_dataset.py

```diff
@@ -1179,63 +1181,73 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         # Here `del` is used to del the pyarrow tables. This properly closes the files used for memory mapped tables
         self.__del__()
 
-    def save_to_disk(self, dataset_path: str, fs=None):
+    def save_to_disk(
+        self,
+        dataset_path: PathLike,
+        fs="deprecated",
+        num_shards: Optional[int] = None,
```

Comment:
If we want to be consistent with the other methods (`push_to_hub`, `download_and_prepare`), this param should be renamed to `max_shard_size`.

Reply:
Nope, this parameter is a number of shard, not the maximum size of every shard :)

For training you often need your dataset to be sharded in a number of files that is a factor of the number of nodes you have - so you need to be able to specify the exact number of shards

Reply:
> Nope, this parameter is a number of shard, not the maximum size of every shard :)

What I meant is renaming + updating the logic.

> For training you often need your dataset to be sharded in a number of files that is a factor of the number of nodes you have - so you need to be able to specify the exact number of shards

For this particular case, one could use `shard` + `save_to_disk`, no? When it comes to the API design, I'm not sure doing this at once in `save_to_disk` is a good idea

Reply:
> For this particular case, one could use shard + save_to_disk, no?

`save_to_disk` requires one directory per dataset, as a sort of bundle. The idea is to end up with one directory containing all the shards, named automatically with the shard idx. Moreover both saving and reloading take one line: it saves some boilerplate code

> When it comes to the API design, I'm not sure doing this at once in save_to_disk is a good idea

What should the user journey look like in that case ? In Dask you can repartition a dataset and then call to_parquet. In Apache beam you can pass num_shards to the WriteTo* functions.

Reply:
As I said, my main concern here is consistency with the rest of the API, i.e., I would rather have a bit of boilerplate (like 3-4 lines of code in this particular case) and consistent API than no boilerplate but inconsistent API (or more ways to do the same thing).

> What should the user journey look like in that case ? In Dask you can repartition a dataset and then call to_parquet. In Apache beam you can pass num_shards to the WriteTo* functions.

We can do several things:
* we should clarify how to use Datasets for distributed training in the docs (shard, ...).
* considering `save_to_disk` supports writing to remote filesystems, we should allow the same in the `to_<format>` methods (e.g. to enable writing Parquet files to GCS)
* advanced usage: support `fsspec` paths as `cache_file_name` to allow writing cached (arrow) files directly to remote filesystems (e.g. `ds.shard(..., cache_file_name="gs://")`) 

And in this PR,  instead of `num_shards`, we could add `max_shard_size` to `save_to_disk` to have the ability to reduce the size of the written Arrow files and leverage multiprocessing (IIRC on some platforms loading large files can be problematic)

Reply:
> As I said, my main concern here is consistency with the rest of the API, i.e., I would rather have a bit of boilerplate (like 3-4 lines of code in this particular case) and consistent API than no boilerplate but inconsistent API (or more ways to do the same thing).

Ok I see, but in this case the boiler plate code is not great IMO. Either it uses `save_to_disk` and we end up with tons of directories, or we use `.map(, cache_file_name=...)` and it feels weird. Maybe we can just add `to_arrow` ?

> we should clarify how to use Datasets for distributed training in the docs (shard, ...).

+ 1 :)

> considering save_to_disk supports writing to remote filesystems, we should allow the same in the to_<format> methods (e.g. to enable writing Parquet files to GCS)

+1 as well !

> advanced usage: support fsspec paths as cache_file_name to allow writing cached (arrow) files directly to remote filesystems (e.g. ds.shard(..., cache_file_name="gs://"))

not 100% sure, since `.shard` creates a new dataset - it doesn't write the dataset on disk. It may write an indices mapping but that's it

Reply:
> Ok I see, but in this case the boiler plate code is not great IMO. Either it uses save_to_disk and we end up with tons of directories, or we use .map(, cache_file_name=...) and it feels weird. Maybe we can just add to_arrow ?

`map` is needed to embed audio/image bytes, right? I think we should release version 3.0 soon (at the beginning of next year?) and start embedding them as we generate a dataset. And then the boilerplate would become much simpler.

> not 100% sure, since .shard creates a new dataset - it doesn't write the dataset on disk. It may write an indices mapping but that's it

Oh, you are right! We could either raise an error in this case or make sure a new dataset is generated (and only do optimizations if `cache_file_name` is written locally).

Reply:
> map is needed to embed audio/image bytes, right?

yea it keeps the links to files by default

> and start embedding them as we generate a dataset

that'b be nice, and we can check with audio folks to make sure it doesn't break stuff ^^

So tin the end o save a dataset as sharded Arrow file here are some options:

1. this PR: add `num_shards` to `save_to_disk`
```python
ds.save_to_disk("output_dir", num_shards=128)
# later
ds = load_from_disk("output_dir")
```
2. add `to_arrow`
```python
num_shards = 128
os.makedirs("output_dir")
for shard_idx in tqdm(range(num_shards), unit="shard"):
    ds.shard(
        num_shards=num_shards,
        index=shard_idx,
        contiguous=True
    ).to_arrow(f"output_dir/data_{shard_idx}.arrow")
# later
ds = concatenate_datasets([
    Dataset.from_file(f"output_dir/data_{shard_idx}.arrow"")
    for shard_idx in range(num_shards)
])
```
3. use map
```python
num_shards = 128
os.makedirs("output_dir")
for shard_idx in tqdm(range(num_shards), unit="shard"):
    ds.shard(
        num_shards=num_shards,
        index=shard_idx,
        contiguous=True
    ).map(cache_file_name=f"output_dir/data_{shard_idx}.arrow")
# later
ds = concatenate_datasets([
    Dataset.from_file(f"output_dir/data_{shard_idx}.arrow"")
    for shard_idx in range(num_shards)
])
```

My concerns so far:
1. adds a new parameter `num_shards` that isn't consistent with `push_to_hub` which has `max_shard_sizes`. We could have both `max_shard_sizes` and `num_shards` in `save_to_disk` and `push_to_hub` IMO - they correspond to different use cases. Apache beam also has both options. Dask as well, but named differently.
2. we need to add a new method + maybe the names `to_arrow`/`from_file` are not fully aligned
3. `map` doesn't embed files. And even if you specify a`cache_file_name`, it may end up writing multiple files using a suffix in multiprocessing. In the future I was thinking of sharding `map` by default to avoid large files.
2 & 3. would have been nice to have `contiguous=True` in `.shard()` by default IMO - not sure if we can fix that

Which one do you prefer ? Do you have other ideas ? Personally I think than 1 and 2 are fine (and small pref for 1 ^^)

Reply:
I like the idea of having both `num_shards` and `max_shard_size` in `push_to_hub`/`save_to_disk` to ensure consistency, so I vote for option 1.